### PR TITLE
store: call RecordWrite() before graphDriver Cleanup()

### PR DESCRIPTION
**Pull request by giuseppe**
Move the execution of RecordWrite() before the graphDriver Cleanup(). This addresses a longstanding issue that occurs when the Podman cleanup process is forcely terminated and on some occasions the termination happens after the Cleanup() but before the change is recorded.  This causes that the next user is not notified about the change and will mount the container without the home directory below (the infamous /var/lib/containers/storage/overlay mount). Then when the next time the graphDriver is initialized, the home directory is mounted on top of the existing mounts causing some containers to fail with ENOENT since all files are hidden and some others cannot be cleaned up since their mount directory is covered by the home directory mount.

Closes: https://github.com/containers/podman/issues/18831
Closes: https://github.com/containers/podman/issues/17216
Closes: https://github.com/containers/podman/issues/17042

### CODE REVIEW

**Comment by giuseppe** [comment]
@edsantiago I am quite sure this solves the issue we've seen.

It is marked as Draft, as I am still testing it though to get more confidence,

**Comment by giuseppe** [comment]
@mtrmac @rhatdan @nalind PTAL

**Comment by rhatdan** [comment]
LGTM

**Comment by giuseppe** [comment]
@vrothberg @flouthoc PTAL

**Review by flouthoc — APPROVED** [review]
LGTM

**Review by vrothberg — COMMENTED** [review]
Amazing job tracking this down, @giuseppe !

**Comment by rhatdan** [comment]
Lets open a PR to get this into Podman.

**Comment by giuseppe** [comment]
here it is: https://github.com/containers/podman/pull/20273

File: store.go [hunk 1]
```diff
@@ -3407,16 +3407,16 @@ func (s *store) Shutdown(force bool) ([]string, error) {
...
 		// We don’t retain the lastWrite value, and treat this update as if someone else did the .Cleanup(),
 		// so that we reload after a .Shutdown() the same way other processes would.
 		// Shutdown() is basically an error path, so reliability is more important than performance.
 		if _, err2 := s.graphLock.RecordWrite(); err2 != nil {
-			if err == nil {
-				err = err2
-			} else {
-				err = fmt.Errorf("(graphLock.RecordWrite failed: %v) %w", err2, err)
-			}
+			err = fmt.Errorf("(graphLock.RecordWrite failed: %w", err2)
```
Comment (nalind) [inline store.go:3414]: Mismatched parentheses in the format string.

Comment (giuseppe) [reply store.go:3414]: thanks for spotting it, opened a PR: https://github.com/containers/storage/pull/1726

**Pull request merged**